### ISSUE TITLE
Adding provisioned concurrency to AWS Lambda results in temporary "Internal server error" 

### ISSUE THREAD

**Issue by richardkazuomiller** [serverless/serverless#8039]
I have a lambda function that did not have any provisioned concurrency. When I added `provisionedConcurrency: 10` to my `serverless.yml` file and deployed, my function resulted in a "Internal server error" response. API Gateway showed the following error: 

```
The API with ID {id} doesn’t include a resource with path /* having an integration {function_arn} on the ANY method.
```

After the deploy was finished and the provisioned concurrency was ready, the error stopped. If I change the provisioned concurrency from 10 to 5, for example, I don't receive any error, so the issue seems to be related to adding provisioned capacity to a function that didn't already have it.

<details>
<summary><code>serverless.yml</code></summary>

```yaml
provider:
  name: aws
  runtime: nodejs12.x
  apiGateway:
    binaryMediaTypes:
      - '*/*'
  stage: dev
  region: ap-northeast-1
functions:
  app:
    handler: index.handler
    timeout: 10
    events:
      - http: ANY /
      - http: 'ANY {proxy+}'
    provisionedConcurrency: 10
```

</details>

<details>
<summary><b><code>⚠️⚠️ REPLACE WITH FULL COMMAND NAME</code> output</b></summary>

```
> serverless deploy -v

Serverless: Packaging service...
Serverless: Excluding development dependencies...
Serverless: Uploading CloudFormation file to S3...
Serverless: Uploading artifacts...
Serverless: Uploading service {my_app}.zip file to S3 (68.49 MB)...
Serverless: Validating template...
Serverless: Updating Stack...
Serverless: Checking Stack update progress...
CloudFormation - UPDATE_IN_PROGRESS - AWS::CloudFormation::Stack - {my_app}
CloudFormation - UPDATE_IN_PROGRESS - AWS::Lambda::Function - AppLambdaFunction
CloudFormation - UPDATE_COMPLETE - AWS::Lambda::Function - AppLambdaFunction
CloudFormation - UPDATE_IN_PROGRESS - AWS::ApiGateway::Method - ApiGatewayMethodAny
CloudFormation - CREATE_IN_PROGRESS - AWS::Lambda::Alias - AppProvConcLambdaAlias
CloudFormation - UPDATE_IN_PROGRESS - AWS::ApiGateway::Method - ApiGatewayMethodProxyVarAny
CloudFormation - CREATE_IN_PROGRESS - AWS::Lambda::Alias - AppProvConcLambdaAlias
CloudFormation - UPDATE_COMPLETE - AWS::ApiGateway::Method - ApiGatewayMethodAny
CloudFormation - UPDATE_COMPLETE - AWS::ApiGateway::Method - ApiGatewayMethodProxyVarAny
CloudFormation - CREATE_IN_PROGRESS - AWS::ApiGateway::Deployment - ApiGatewayDeployment1596423600981
CloudFormation - CREATE_IN_PROGRESS - AWS::ApiGateway::Deployment - ApiGatewayDeployment1596423600981
CloudFormation - CREATE_COMPLETE - AWS::ApiGateway::Deployment - ApiGatewayDeployment1596423600981
CloudFormation - CREATE_COMPLETE - AWS::Lambda::Alias - AppProvConcLambdaAlias
CloudFormation - UPDATE_IN_PROGRESS - AWS::Lambda::Permission - AppLambdaPermissionApiGateway
CloudFormation - UPDATE_IN_PROGRESS - AWS::Lambda::Permission - AppLambdaPermissionApiGateway
CloudFormation - UPDATE_COMPLETE - AWS::Lambda::Permission - AppLambdaPermissionApiGateway
CloudFormation - UPDATE_COMPLETE_CLEANUP_IN_PROGRESS - AWS::CloudFormation::Stack - {my_app}
CloudFormation - DELETE_IN_PROGRESS - AWS::Lambda::Permission - AppLambdaPermissionApiGateway
CloudFormation - DELETE_IN_PROGRESS - AWS::ApiGateway::Deployment - ApiGatewayDeployment1596423456026
CloudFormation - DELETE_COMPLETE - AWS::ApiGateway::Deployment - ApiGatewayDeployment1596423456026
CloudFormation - DELETE_COMPLETE - AWS::Lambda::Permission - AppLambdaPermissionApiGateway
CloudFormation - UPDATE_COMPLETE - AWS::CloudFormation::Stack - {my_app}
```

</details>

<b>Installed version</b>

```
Framework Core: 1.67.0
Plugin: 3.6.1
SDK: 2.3.0
Components: 2.22.3
```

**Comment by medikoo** [comment]
@richardkazuomiller it might be an issue on AWS side (e.g. they need some propagation time, through which your endpoint is not accessible). Have you confirmed what exactly caused `Internal server error`?

**Comment by richardkazuomiller** [comment]
No, I don't know exactly what the cause is, but I do know that removing provisioned concurrency and setting it back to 10 reproduces the error until the provisioned concurrency is ready. My best guess is that it has something to do with the fact that API Gateway doesn't know what function to invoke, based on the error `The API with ID {id} doesn’t include a resource with path /* having an integration {function_arn} on the ANY method.` displayed in the console. There are no invocation logs, which makes sense if API Gateway can't invoke the function.

I noticed that `CloudFormation - UPDATE_IN_PROGRESS - AWS::ApiGateway::Method - ApiGatewayMethodAny` starts before `CREATE_COMPLETE - AWS::Lambda::Alias - AppProvConcLambdaAlias`. Shouldn't the Lambda Alias be created before the API Gateway starts being modified?

**Comment by medikoo** [comment]
@richardkazuomiller this is all the questions you should direct to AWS support. On Framework side we just prepare template and deploy it, rest is in hands of AWS

**Comment by richardkazuomiller** [comment]
So you're 100% certain this is not a bug in CloudFormation and not Serverless? There's no way to configure it to deploy in a way that doesn't cause an error in API Gateway?

**Comment by medikoo** [comment]
> There's no way to configure it to deploy in a way that doesn't cause an error in API Gateway?

I don't know, but I'm pretty sure AWS support will be able to help

**Comment by radove** [comment]
@richardkazuomiller we had issues with Provisioned Concurrency as well.  We forced the version of serverless to an older release, and had success.  We used serverless@1.80.0 for this.  Let me know if rolling back the version, fixes the issue.

**Comment by Edujugon** [comment]
Hi,
We just faced a similar issue. If we add the `provisionedConcurrency` key to a lambda that was already deployed, when hitting the endpoint it gives an `Internal server error` response.

I checked that lambda in the console, and it had 2 API event sources applied whereas before it had only one. Then I moved to the APIGW console, edited the Lambda Function name in the Lambda Proxy integration section, redeployed the API (from the console), and then it started to work...

**Comment by burketyler** [comment]
We had this issue occur during a production deployment, which was completely unexpected as you can perform the same action in the console with no downtime. I haven't traced the exact series of events that leads to this occurring but it appears to be a timing issue during the deploy where either 1) API Gateway has doesn't yet have permissions to invoke the newly created alias, or 2) the lambda code has not been completely deployed before API Gateway has been switched to route to the new alias.

Regardless of whether this is caused by Serverless or CloudFormation I would have expected an issue that can cause downtime during an `sls deploy` to be properly documented, which doesn't appear to be the case. How has something like this not been address in over a year?

**Comment by pgrzesik** [comment]
Hello @burketyler - as you can see from the number of reports under this issue, it was not very common problem.  Do you have the ability to provide a reproducible example of your specific issue?  What version of the Framework are you using?

**Comment by wolfejw86** [comment]
Looks like there are several gotchas related to updating a lambda with provisioned concurrency. While not about serverless framework, I found this article helpful since I have just now run into the above issue as well. It helped me understand the pieces at play better - posting here in case it helps someone else find their zero downtime solution faster.

https://futurice.com/blog/building-an-aws-serverless-system-provisioned-concurrency-zero-downtime
